File: browser-api/src/main/java/org/jetbrains/bio/browser/desktop/MainPanel.java
package org.jetbrains.bio.browser.desktop;

import org.jdesktop.swingx.util.OS;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.bio.browser.GenomeBrowser;
import org.jetbrains.bio.browser.model.BrowserModel;
import org.jetbrains.bio.browser.tracks.TrackView;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.util.HashMap;
import java.util.List;
import java.util.function.Consumer;

/**
 * Created on 13/06/13
 * Author: Sergey Dmitriev
 */
public class MainPanel extends JPanel implements ActionListener {

  private final HashMap<String, Consumer<ActionEvent>> myActionsMap = new HashMap<>();
  private final TrackListComponent myTrackListComponent;
  private final ControlsPanel myControlPane;
  private final ScrollBarModel myScrollBarModel;

  public MainPanel(final DesktopGenomeBrowser browser, final List<TrackView> trackViews) {
    setLayout(new BorderLayout());

    // controls
    myControlPane = new ControlsPanel(browser);
    add(myControlPane, BorderLayout.NORTH);

    myTrackListComponent = new TrackListComponent(this, browser, trackViews);
    add(myTrackListComponent.getContentPane(), BorderLayout.CENTER);

    // Genome region scroll bar
    final JScrollBar scrollBar = new JScrollBar(JScrollBar.HORIZONTAL);
    myScrollBarModel = new ScrollBarModel(browser.getBrowserModel());
    scrollBar.setModel(myScrollBarModel);
    add(scrollBar, BorderLayout.SOUTH);
  }

  public void setModel(final BrowserModel browserModel) {
    myControlPane.setBrowserModel(browserModel);
    myScrollBarModel.setBrowserModel(browserModel);
    myTrackListComponent.getContentPane().setColumnHeaderView(GenomeBrowser.Companion.createHeaderView(browserModel));
  }

  public TrackListComponent getTrackListComponent() {
    return myTrackListComponent;
  }

  @Override
  public void actionPerformed(@NotNull final ActionEvent e) {
    final String actionCommand = e.getActionCommand();
    final Consumer<ActionEvent> handler = myActionsMap.get(actionCommand);
    if (handler != null) {
      handler.accept(e);
    }
  }

  public void fillMenu(final JMenuBar mainMenu) {
    final TrackListController controller = myTrackListComponent.getTrackListController();
    // Navigation
    final JMenu navigationMenu = new JMenu("Navigation");
    navigationMenu.setMnemonic('n');
    mainMenu.add(navigationMenu);
    registerMenuItem(OS.isMacOSX() ? KeyEvent.VK_OPEN_BRACKET : KeyEvent.VK_LEFT,
                     OS.isMacOSX() ? ActionEvent.META_MASK : ActionEvent.ALT_MASK,
                     e -> controller.undo(), navigationMenu, "Back", 0, false, false);
    registerMenuItem(OS.isMacOSX() ? KeyEvent.VK_CLOSE_BRACKET : KeyEvent.VK_RIGHT,
                     OS.isMacOSX() ? ActionEvent.META_MASK : ActionEvent.ALT_MASK,
                     e -> controller.redo(), navigationMenu, "Forward", 0, false, false);

    navigationMenu.addSeparator();
    registerMenuItem(KeyEvent.VK_LEFT, 0, e -> controller.doScrollTrack(true, false), navigationMenu,
                     "Scroll Left", (int) 'l', false, false);
    registerMenuItem(KeyEvent.VK_RIGHT, 0, e -> controller.doScrollTrack(false, false), navigationMenu,
                     "Scroll Right", (int) 'r', false, false);
    registerMenuItem(KeyEvent.VK_LEFT, getMetaModifier(), e -> controller.doScrollTrack(true, true), navigationMenu,
                     "Scroll Window Left", 0, false, false);
    registerMenuItem(KeyEvent.VK_RIGHT, getMetaModifier(), e -> controller.doScrollTrack(false, true), navigationMenu,
                     "Scroll Window Right", 0, false, false);

    // Zoom actions
    navigationMenu.addSeparator();

    mainMenu.add(navigationMenu);
    registerMenuItem(KeyEvent.VK_EQUALS, 0, e -> controller.doZoom2x(true), navigationMenu,
                     "Zoom In", (int) 'i', false, false);
    registerMenuItem(KeyEvent.VK_MINUS, 0, e -> controller.doZoom2x(false), navigationMenu,
                     "Zoom Out", (int) 'o', false, false);

    // Misc
    final JMenu miscMenu = new JMenu("Misc");
    miscMenu.setMnemonic('m');
    mainMenu.add(miscMenu);
    registerMenuItem(KeyEvent.VK_A, getMetaModifier(), e -> controller.doSelectAllTracks(), miscMenu,
                     "Select All Tracks", (int) 'a', false, false);
    miscMenu.addSeparator();
    registerMenuItem(KeyEvent.VK_S, getMetaModifier(), e -> myTrackListComponent.doTakeScreenshot(false), miscMenu,
                     "Screenshot", (int) 's', false, false);
    registerMenuItem(KeyEvent.VK_S, getMetaModifier() | ActionEvent.SHIFT_MASK,
                     e -> myTrackListComponent.doTakeScreenshot(true), miscMenu,
                     "Screenshot Selected Track", 0, false, false);
    miscMenu.addSeparator();
    registerMenuItem(null, 0, e -> controller.doToggleLegendVisibility(), miscMenu, "Show legend", (int) 'l',
                     true, myTrackListComponent.getUIModel().get(TrackView.SHOW_LEGEND));
    registerMenuItem(null, 0, e -> controller.doToggleAxisVisibility(), miscMenu, "Show axis", (int) 'x',
                     true, myTrackListComponent.getUIModel().get(TrackView.SHOW_AXIS));
  }

  public static int getMetaModifier() {
    return OS.isMacOSX() ? ActionEvent.META_MASK : ActionEvent.CTRL_MASK;
  }

  private void registerMenuItem(@Nullable final Integer keyChar,
                                     final int modifiers,
                                     final Consumer<ActionEvent> handler,
                                     final JMenu menu,
                                     final String title,
                                     final int mnemonic,
                                     final boolean isCheckBox,
                                     final boolean checkBoxDefaultState) {
    final JMenuItem menuItem = isCheckBox
                               ? new JCheckBoxMenuItem(title, checkBoxDefaultState)
                               : new JMenuItem(title);
    menuItem.setMnemonic(mnemonic);
    menu.add(menuItem);
    if (keyChar != null) {
      menuItem.setAccelerator(KeyStroke.getKeyStroke(keyChar, modifiers));
    }
    menuItem.addActionListener(this);
    myActionsMap.put(title, handler);
  }
}

File: browser-api/src/main/kotlin/org/jetbrains/bio/browser/desktop/MainPanel.kt
package org.jetbrains.bio.browser.desktop

import org.jdesktop.swingx.util.OS.isMacOSX
import org.jetbrains.bio.browser.GenomeBrowser
import org.jetbrains.bio.browser.model.BrowserModel
import org.jetbrains.bio.browser.tracks.TrackView
import java.awt.BorderLayout
import javax.swing.*
import javax.swing.KeyStroke.getKeyStroke

/**
 * Created on 13/06/13
 * Author: Sergey Dmitriev
 */
class MainPanel(browser: DesktopGenomeBrowser, trackViews: List<TrackView>) : JPanel() {
    val trackListComponent = TrackListComponent(this, browser, trackViews)

    private val searchPanel = SearchPanel(browser)
    private val scrollBarModel = ScrollBarModel(browser.model)

    init {
        layout = BorderLayout()

        add(searchPanel, BorderLayout.NORTH)
        add(trackListComponent.contentPane, BorderLayout.CENTER)

        val scrollBar = JScrollBar(JScrollBar.HORIZONTAL)
        scrollBar.model = scrollBarModel
        add(scrollBar, BorderLayout.SOUTH)
    }

    fun setModel(browserModel: BrowserModel) {
        searchPanel.setBrowserModel(browserModel)
        scrollBarModel.setBrowserModel(browserModel)
        trackListComponent.contentPane.setColumnHeaderView(GenomeBrowser.createHeaderView(browserModel))
    }

    fun fillMenu(mainMenu: JMenuBar) {
        val controller = trackListComponent.trackListController
        val navigationMenu = NavigationMenu().apply {
            back.addActionListener { controller.undo() }
            forward.addActionListener { controller.redo() }

            scrollLeft.addActionListener { controller.doScrollTrack(true, false) }
            scrollRight.addActionListener { controller.doScrollTrack(false, false) }
            scrollWindowLeft.addActionListener { controller.doScrollTrack(true, true) }
            scrollWindowRight.addActionListener { controller.doScrollTrack(false, true) }

            zoomIn.addActionListener { controller.doZoom2x(true) }
            zoomOut.addActionListener { controller.doZoom2x(false) }
        }

        val miscMenu = MiscMenu().apply {
            screenshot.addActionListener { trackListComponent.doTakeScreenshot(false) }
            screenshotSelected.addActionListener { trackListComponent.doTakeScreenshot(true) }

            showLegend.state = trackListComponent.uiModel[TrackView.SHOW_LEGEND]
            showLegend.addActionListener { controller.doToggleLegendVisibility() }
            showAxis.state = trackListComponent.uiModel[TrackView.SHOW_AXIS]
            showAxis.addActionListener { controller.doToggleAxisVisibility() }
        }

        mainMenu.add(navigationMenu)
        mainMenu.add(miscMenu)
    }
}

class NavigationMenu : JMenu("Navigation") {
    val back = JMenuItem("Back").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta OPEN_BRACKET" else "alt LEFT")
    }
    val forward = JMenuItem("Forward").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta CLOSE_BRACKET" else "alt RIGHT")
    }

    val scrollLeft = JMenuItem("Scroll Left").apply {
        accelerator = getKeyStroke("LEFT")
    }
    val scrollRight = JMenuItem("Scroll Right").apply {
        accelerator = getKeyStroke("RIGHT")
    }
    val scrollWindowLeft = JMenuItem("Scroll Window Left").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta LEFT" else "ctrl LEFT")
    }
    val scrollWindowRight = JMenuItem("Scroll Window Right").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta RIGHT" else "ctrl RIGHT")
    }

    val zoomIn = JMenuItem("Zoom In").apply {
        accelerator = getKeyStroke('=')
    }
    val zoomOut = JMenuItem("Zoom Out").apply {
        accelerator = getKeyStroke('-')
    }

    init {
        add(back)
        add(forward)
        addSeparator()
        add(scrollLeft)
        add(scrollRight)
        add(scrollWindowLeft)
        add(scrollWindowRight)
        addSeparator()
        add(zoomIn)
        add(zoomOut)
    }
}

class MiscMenu : JMenu("Misc") {
    val screenshot = JMenuItem("Screenshot").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta S" else "ctrl S")
    }
    val screenshotSelected = JMenuItem("Screenshot Selected Track").apply {
        accelerator = getKeyStroke(if (isMacOSX()) "meta shift S" else "ctrl shift S")
    }

    val showLegend = JCheckBoxMenuItem("Show Legend")
    val showAxis = JCheckBoxMenuItem("Show Axis")

    init {
        add(screenshot)
        add(screenshotSelected)
        addSeparator()
        add(showLegend)
        add(showAxis)
    }
}
